nutils/types.py

```diff
@@ -1492,6 +1492,56 @@ def del_lru(self):
       lru = self._sorted_keys.pop()
     del self[lru]
 
+def lru_cache(func=None, maxsize=None):
+  '''Buffer-aware LRU cache.
+
+  Returns values from a cache for previously seen arguments. Arguments must be
+  hasheable objects or Numpy arrays, with the caveat that writeable arrays
+  silently disable caching. Arrays are equality-tested based on the underlying
+  buffer, with a callback placed on the memory release that removes the
+  corresponding cache entry.
+  '''
+
+  if func is None:
+    return functools.partial(lru_cache, maxsize=maxsize)
+
+  cache = {} if maxsize is None else lru_dict(maxsize)
+
+  @functools.wraps(func)
+  def wrapped(*args):
+    key = []
+    bases = []
+    for arg in args:
+      if isinstance(arg, numpy.ndarray):
+        obj = arg
+        while True:
```

Comment:
I guess we could also introduce a generator to traverse the bases in `numerics.py`
```python
def bases(obj, least=False):
  while isinstance(obj, numpy.ndarray):
    yield obj
    obj = obj.base
    if isinstance(obj, numpy.lib.stride_tricks.DummyArray):
      obj = obj.base
  if least:
    yield obj
```
Then simplify the code to
```python
obj = arg
for obj in bases(arg, least=True): # least = True so obj point to last nonnil after loop
  if isinstance(obj, arraydata): break
  if obj.flags.writeable:
    return func(*args)
```

Notice the iterator would also simplify `numeric.freeze`
```python
def freeze(arg):
  assert isinstance(arg, numpy.ndarray)
  for a in bases(arg):
    a.flags.writeable = False
  return arg
```

Appendix:
the base hierarchy in regular expression
```ocaml
let N = ndarray
    D = DummyArray
    W = arraydata
    N = None
in A(A|DA)*(W|N)
```

Reply:
I had this decorator at one stage, but at present there are only two functions that traverse bases: `types._isimmutable`, which is recursive by nature, and `numeric.freeze` - so there is not that much gained by the additional abstraction. Moreover, since `numeric` imports `types`, the base iterator should be placed in the latter to be used by `_isimmutable`, which is an odd location. All in all I prefer to kick this particular can a little down the road.

Reply:
PS. I'm not sure that I understand the purpose of the regular expression.

Reply:
I didn't notice `_isimutable` but here in `lru_cache` we also traverse the bases.
The regex is just for documentation purpose.

Reply:
Fair point. In the latest (and hopefully final) version I removed `numeric.freeze`, instead keeping for that purpose `types.frozenarray` which has the benefit of touching fewer code and being backwards compatible. Supporting `types.frozenarray`, `types._isimmutable` and `types.lru_cache` there is now `types._array_bases`.